src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -306,6 +306,21 @@ jlong ObjectSynchronizer::_last_async_deflation_time_ns = 0;
 static uintx _no_progress_cnt = 0;
 static bool _no_progress_skip_increment = false;
 
+#define CHECK_THROW_NOSYNC_IE(obj) \
+  if (obj->klass()->is_inline_klass()) { \
+    ResourceMark rm(THREAD); \
+    const char* desc = "Cannot synchronize on an instance of value class "; \
+    const char* className = obj->klass()->external_name(); \
+    size_t msglen = strlen(desc) + strlen(className) + 1; \
+    char* message = NEW_RESOURCE_ARRAY(char, msglen); \
+    if (nullptr == message) { \
+      THROW_MSG(vmSymbols::java_lang_IdentityException(), className); \
+    } else { \
+      jio_snprintf(message, msglen, "%s%s", desc, className); \
+      THROW_MSG(vmSymbols::java_lang_IdentityException(), message); \
+    } \
+  }
+
 #define CHECK_THROW_NOSYNC_IMSE(obj)  \
```

Comment:
I added a short comment why this is needed. like:

// These checks are required for wait, notify and exit to avoid inflating the monitor to
// find out this inline type object cannot be locked.